src/coreclr/jit/inlinepolicy.cpp

```diff
@@ -1440,6 +1440,60 @@ void ExtendedDefaultPolicy::NoteDouble(InlineObservation obs, double value)
     m_ProfileFrequency = value;
 }
 
+//------------------------------------------------------------------------
+// BudgetCheck: see if this inline would exceed the current budget
+//
+// Returns:
+//   True if inline would exceed the budget.
+//
+bool ExtendedDefaultPolicy::BudgetCheck() const
+{
+    if (m_IsPrejitRoot)
+    {
+        // Only relevant if we're actually inlining.
+        return false;
+    }
+
+    // The strategy tracks the amout of inlining done so far, so it performs the actual check.
+    const bool overBudget = m_RootCompiler->m_inlineStrategy->BudgetCheck(m_CodeSize);
+
+    if (overBudget)
+    {
+        assert(m_IsForceInlineKnown);
+        assert(m_CallsiteDepth > 0);
+
+        if (m_IsForceInline)
+        {
+            if (m_CallsiteDepth == 1)
+            {
+                // If the candidate is a forceinline and the callsite is
+                // not too deep, allow the inline even if it goes over budget.
```

Comment:
Does this mean inline even if the callee is very large or was size eligibility already determined?

Reply:
Yes, but that is our current behavior and I didn't want to change that (to avoid unexpected regressions):
- Inline AggressiveInlining of any size if they're top-level (root) in the call-graph
- Apply time-budget to all non top-level ones.

If someone marks a huge method as AggressiveInlining they're likely their own worst enemy 🙂 We've seen a lot of those and not sure what motivates people to do that exactly, hopefully, they do that with benchmark measurements.